### DISCUSSION TITLE
How to initialize reference vs. initial porosity.

### DISCUSSION THREAD

**Discussion by homel1** [GEOS-DEV/GEOS#1594] [Infrastructure]
For many constitutive models it is simpler to define a "Reference" state to be at some reference pressure, temp, etc.

For porosity, this suggests making a user-specified reference porosity to be the porosity at some reference pressure (e.g. at atmospheric pressure).  This is complicated for porosity in a reservoir, because there is spatially varying pressure.  For the user it is desirable to be able to specify the initial porosity at the equilibrium pressure distribution.

The discussion is whether this user-specified value should be the "Reference_porosity" used in the porosity update function (simple solution) or whether to give the user an option to specify initial porosity, where there would be some internal function compute_reference_porosity_from_initial

One challenge is in defining the initial equilibrium state. 

Assuming a material has a prescribed crush/poro-elastic response: porosity(reference_porosity, pressure), where density of the porous material depends on solid density and porosity, and the pressure distribution with depth depends on density and gravity, the user-specified initial porosity distribution will generally not be in equilibrium, if the reference porosity doesn't depend on depth.

However, the compute_reference_porosity_from_initial function can be defined based on an initial pressure distribution (consistent with the prescribed initial porosity/density field) to find a reference porosity that makes the initial state be in equilibrium.

1. The user specifies: 
initial_porosity as a function of depth
density of the solid phase
2. The code computes: density as a function of depth
3. The code computes: equilibrium pressure as a function of depth
3. The code computes: reference porosity at each element as a function of the equilibrium pressure and prescribed initial porosity.

The advantages of this approach are (i) the constitutive model describes a material response independent of the application, (ii) the porosity model is forced to be in equilibrium with the initial conditions.

This requires that reference_porosity be a state variable (changing element-to-element) not a parameter that is the same for all elements in the element region.

**Comment by TotoGaz** [comment]
In a nutshell, you are looking for a "_reference porosity provider_", am I correct?
As such, would it make sense to let the user let make his own choice in the interface?
* If he wants a uniform reference porosity, implement a `UniformReferencePorosityProvider`.
* In case it depends of depth only, make some `DepthReferencePorosityProvider`.
* If you want the reference porosity to be the result of a computation, would it make sense to implement some `ComputationalPorosityProvider` with all the arguments needed (`initial_porosity`, numerical parameters).

There may be multiple recipes to perform such a task, so you may need to provide multiple implementations.
That way you impose nothing to the user and it's always possible to add another solution for a dedicated need.

Does this help you?
